llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp

```diff
@@ -0,0 +1,294 @@
+
+private:
+  SPIRITTAnnotationsPass Impl;
+};
+
+} // namespace
+
+char SPIRITTAnnotationsLegacyPass::ID = 0;
+INITIALIZE_PASS(SPIRITTAnnotationsLegacyPass, "SPIRITTAnnotations",
+                "Insert ITT annotations in SPIR code", false, false)
+
+// Public interface to the SPIRITTAnnotationsPass.
+ModulePass *llvm::createSPIRITTAnnotationsPass() {
+  return new SPIRITTAnnotationsLegacyPass();
+}
+
+namespace {
+
+// Check for calling convention of a function.
+bool isSPIRKernel(Function &F) {
+  return F.getCallingConv() == CallingConv::SPIR_KERNEL;
+}
+
+Instruction *emitCall(Module &M, Type *RetTy, StringRef FunctionName,
+                      ArrayRef<Value *> Args, Instruction *InsertBefore) {
+  SmallVector<Type *, 8> ArgTys(Args.size());
+  for (unsigned I = 0; I < Args.size(); ++I)
+    ArgTys[I] = Args[I]->getType();
+  auto *FT = FunctionType::get(RetTy, ArgTys, false /*isVarArg*/);
+  FunctionCallee FC = M.getOrInsertFunction(FunctionName, FT);
+  assert(FC.getCallee() && "Instruction creation failed");
+  auto *Call = CallInst::Create(FT, FC.getCallee(), Args, "", InsertBefore);
+  return Call;
+}
+
+// Insert instrumental annotation calls, that has no arguments (for example
+// work items start/finish/resume and barrier annotation.
+bool insertSimpleInstrumentationCall(Module &M, StringRef Name,
+                                     Instruction *Position) {
+  Type *VoidTy = Type::getVoidTy(M.getContext());
+  ArrayRef<Value *> Args;
+  Instruction *InstrumentationCall = emitCall(M, VoidTy, Name, Args, Position);
+  assert(InstrumentationCall && "Instrumentation call creation failed");
+  return true;
+}
+
+// Insert instrumental annotation calls for SPIR-V atomics.
+bool insertAtomicInstrumentationCall(Module &M, StringRef Name,
+                                     CallInst *AtomicFun,
+                                     Instruction *Position) {
+  LLVMContext &Ctx = M.getContext();
+  Type *VoidTy = Type::getVoidTy(Ctx);
+  Type *Int32Ty = Type::getInt32Ty(Ctx);
+  // __spirv_Atomic... instructions have following arguments:
+  // Pointer, Memory Scope, Memory Semantics and others. To construct Atomic
+  // annotation instructions we need Pointer and Memory Semantic arguments
+  // taken from the original Atomic instruction.
+  Value *Ptr = dyn_cast<Value>(AtomicFun->getArgOperand(0));
+  StringRef AtomicName = AtomicFun->getCalledFunction()->getName();
+  Value *AtomicOp;
+  // Second parameter of Atomic Start/Finish annotation is an Op code of
+  // the instruction, encoded into a value of enum, defined like this on user's/
+  // profiler's side:
+  // enum __itt_atomic_mem_op_t
+  // {
+  //   __itt_mem_load = 0,
+  //   __itt_mem_store = 1,
+  //   __itt_mem_update = 2
+  // }
+  if (AtomicName.contains(SPIRV_ATOMIC_LOAD))
+    AtomicOp = ConstantInt::get(Int32Ty, 0);
+  else if (AtomicName.contains(SPIRV_ATOMIC_STORE))
+    AtomicOp = ConstantInt::get(Int32Ty, 1);
+  else
+    AtomicOp = ConstantInt::get(Int32Ty, 2);
```

Comment:
If you change your prefixes, you could use [`llvm::StringSwitch`](https://llvm.org/doxygen/classllvm_1_1StringSwitch.html) here: I mean use `_Z[number]__spirv_AtomicLoad` and `llvm::StringSwitch::StartsWith`